<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

<h1>Tabla de Contenidos<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introducción-a-Pandas" data-toc-modified-id="Introducción-a-Pandas-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introducción a Pandas</a></span><ul class="toc-item"><li><span><a href="#¿Qué-es-Pandas-y-por-qué-es-importante?" data-toc-modified-id="¿Qué-es-Pandas-y-por-qué-es-importante?-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>¿Qué es Pandas y por qué es importante?</a></span></li></ul></li><li><span><a href="#Series-de-Pandas" data-toc-modified-id="Series-de-Pandas-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Series de Pandas</a></span><ul class="toc-item"><li><span><a href="#¿Qué-es-una-Serie?" data-toc-modified-id="¿Qué-es-una-Serie?-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>¿Qué es una Serie?</a></span></li><li><span><a href="#Creación-de-Series" data-toc-modified-id="Creación-de-Series-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Creación de Series</a></span></li><li><span><a href="#Propiedades-de-las-Series" data-toc-modified-id="Propiedades-de-las-Series-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Propiedades de las Series</a></span></li><li><span><a href="#Indexación-de-las-Series" data-toc-modified-id="Indexación-de-las-Series-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Indexación de las Series</a></span></li></ul></li><li><span><a href="#DataFrames-en-Pandas" data-toc-modified-id="DataFrames-en-Pandas-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>DataFrames en Pandas</a></span><ul class="toc-item"><li><span><a href="#¿Qué-es-un-DataFrame?" data-toc-modified-id="¿Qué-es-un-DataFrame?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>¿Qué es un DataFrame?</a></span></li><li><span><a href="#Creación-de-DataFrames" data-toc-modified-id="Creación-de-DataFrames-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Creación de DataFrames</a></span></li><li><span><a href="#Lectura-de-archivos-en-Pandas" data-toc-modified-id="Lectura-de-archivos-en-Pandas-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Lectura de archivos en Pandas</a></span></li><li><span><a href="#pd.read_csv():" data-toc-modified-id="pd.read_csv():-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span><code>pd.read_csv()</code>:</a></span></li><li><span><a href="#pd.read_excel()" data-toc-modified-id="pd.read_excel()-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span><code>pd.read_excel()</code></a></span></li><li><span><a href="#pd.read_json()" data-toc-modified-id="pd.read_json()-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span><code>pd.read_json()</code></a></span></li><li><span><a href="#pd.read_pickle" data-toc-modified-id="pd.read_pickle-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span><code>pd.read_pickle</code></a></span></li><li><span><a href="#Indexación-en-Pandas-(loc-e-iloc)" data-toc-modified-id="Indexación-en-Pandas-(loc-e-iloc)-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Indexación en Pandas (<code>loc</code> e <code>iloc</code>)</a></span><ul class="toc-item"><li><span><a href="#Selección-a-través-de-un-único-elemento" data-toc-modified-id="Selección-a-través-de-un-único-elemento-3.8.1"><span class="toc-item-num">3.8.1&nbsp;&nbsp;</span>Selección a través de un único elemento</a></span></li><li><span><a href="#Selección-a-través-de-una-lista-de-valores" data-toc-modified-id="Selección-a-través-de-una-lista-de-valores-3.8.2"><span class="toc-item-num">3.8.2&nbsp;&nbsp;</span>Selección a través de una lista de valores</a></span></li><li><span><a href="#Selección-de-un-rango-de-datos-a-través-de-un-corte" data-toc-modified-id="Selección-de-un-rango-de-datos-a-través-de-un-corte-3.8.3"><span class="toc-item-num">3.8.3&nbsp;&nbsp;</span>Selección de un rango de datos a través de un corte</a></span></li><li><span><a href="#Seleccionar-basado-en-una-condición" data-toc-modified-id="Seleccionar-basado-en-una-condición-3.8.4"><span class="toc-item-num">3.8.4&nbsp;&nbsp;</span>Seleccionar basado en una condición</a></span></li></ul></li></ul></li></ul></div>

# Introducción a Pandas

## ¿Qué es Pandas y por qué es importante?

`Pandas` es una biblioteca de Python utilizada para el análisis de datos y la manipulación de estructuras de datos tabulares, es decir, en formato de tablas. Nos ofrece una gran cantidad de funciones y herramientas para trabajar con datos de forma eficiente y efectiva. Algunas de las razones por las que Pandas es importante son:

- Define nuevas estructuras de datos basadas en los **array*s* de la biblioteca NumPy, pero con funcionalidades adicionales. Esto significa que podemos aprovechar las capacidades de NumPy y agregar nuevas funcionalidades específicas para el análisis de datos.
  
- Permite la carga y manipulación de datos de diferentes fuentes, como archivos CSV, Excel, bases de datos SQL, entre otros.
  
- Proporciona funciones para explorar y entender rápidamente la estructura y contenido de los datos.

- Facilita la limpieza y transformación de datos, lo que es fundamental para garantizar la calidad de los datos antes del análisis.

- Ofrece herramientas para realizar análisis estadísticos, visualizaciones y modelado de datos.

- Permite acceder a los datos mediante índices o nombres para filas y columnas. Esto significa que puedes acceder a los datos utilizando etiquetas en lugar de solo posiciones numéricas, lo cual es especialmente útil cuando los datos están etiquetados o indexados de manera significativa.

- Ofrece métodos para reordenar, dividir y combinar conjuntos de datos. Puedes realizar operaciones como clasificación, filtrado y combinación de datos para obtener resultados específicos o realizar análisis más avanzados.

En cuanto a las estructuras de datos en Pandas, existen tres principales:

1. `Series`: Representa una estructura de datos unidimensional similar a un *array* o una columna en una tabla. Cada elemento en una Serie tiene un índice asociado que permite acceder a los datos de forma individual.


2. `DataFrame`: Representa una estructura de datos bidimensional similar a una tabla de una base de datos o una hoja de cálculo de Excel. Consiste en una colección de Series organizadas en columnas. Los *DataFrame*s son especialmente útiles para el análisis de datos tabulares y ofrecen una gran cantidad de funcionalidades para manipular y analizar los datos.


3. `Panel`: Representa una estructura de datos tridimensional, aunque su uso no es tan común como las *Series* y los *DataFrame*s. Puede considerarse como una colección de *DataFrame*s.

Antes de ponernos a trabajar con Pandas, lo primero que tendremos que hacer es instalar la librería usando `pip` o `pip3`: 

```bash
pip install pandas

pip3 install pandas
```

In [6]:
# Tratamiento de datos
# ==============================================================================
import pandas as pd

# Ignorar warnings
# ==============================================================================
import warnings
warnings.filterwarnings("ignore")


# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

# Series de Pandas

## ¿Qué es una Serie?


Como hemos dicho, una **Serie** es una estructura de datos unidimensional en **Pandas**. Se puede pensar en una **Serie** como una columna en una tabla  con etiquetas de índice. Cada elemento en una **Serie** tiene un índice asociado que permite acceder a los datos de forma individual.

Las Series en Pandas tienen las siguientes características:

1. **Datos homogéneos**: Contiene datos de **un solo tipo**, como *integer*s, *float*s, *strings*, etc. 

2. **Índices etiquetados**: Cada elemento en una **Serie** tiene una etiqueta de índice asociada. Los índices pueden ser etiquetas personalizadas o valores numéricos generados automáticamente. Estas etiquetas permiten un acceso más intuitivo a los datos en lugar de depender solo de las posiciones numéricas.

3. **Funcionalidades adicionales**: Proporcionan numerosas funcionalidades para manipular y analizar datos de manera eficiente. Puedes realizar operaciones matemáticas, filtrado de datos, agregación, ordenamiento, alineación de datos, entre otras.

Es importante destacar que las **Series** y los **DataFrames** son los elementos fundamentales en el ecosistema de Pandas, y muchas operaciones de análisis de datos se realizan utilizando estas estructuras de datos.


## Creación de Series

Podemos crear Series de distintas formas: 

- Series vacías.

- Series a partir de un **array**. Recordamos, las **Series** son unidimensionales, por lo tanto solo le podremos pasar **array*s* unidimensionales.

- Series a partir de listas.

- Series a partir de un diccionario.


In [7]:
# creemos algunas Series en Pandas para entenderlas mejor.
# en este primer ejemplo crearemos una Serie vacía

print(" ------ SERIES VACÍAS ------ ")
serie_vacia = pd.Series()
serie_vacia

 ------ SERIES VACÍAS ------ 


Series([], dtype: object)

In [8]:
print(" ------ SERIES a partir de LISTAS ------ ")

# definimos la lista
lista = [109,78, 4, 189, 98]

# convertimos esta lista en una Serie
serie_lista1 = pd.Series(lista)
serie_lista1

 ------ SERIES a partir de LISTAS ------ 


0    109
1     78
2      4
3    189
4     98
dtype: int64

📌 Si nos fijamos el método `pd.Series` nos va a generar automáticamente el índice y lo crea con números consecutivos, empezando por el 0. Sin embargo, también le podemos pasar los índices que queremos usando el parámetro `index`. Tener índices personalizados puede ser útil cuando necesitamos acceder a los elementos de la **Serie** utilizando etiquetas específicas en lugar de las posiciones numéricas predeterminadas. Esto facilita la manipulación y el análisis de los datos. 



> index: Es un parámetro opcional de la función `pd.Series`. En este caso, se proporciona una lista de índices personalizados para la Serie. Los índices indican cómo se etiquetarán los elementos de la Serie. En este ejemplo, los índices se definen en el orden especificado: 3, 4, 5, 6, 7, 8, 9, 1, 2.
>
> Es importante destacar que los índices pueden ser de diferentes tipos, incluidos *integer*s, cadenas, fechas, etc. Además, los índices no necesariamente tienen que estar en orden o ser consecutivos. 

In [9]:
print(" ------ SERIES a partir de LISTAS ESPECIFICANDO ÍNDICE NÚMERICO ------ ")
# vamos a definir una nueva serie pero en este caso vamos a ponerle nuestro propio índice usando el parámetro 'index'
serie_lista2 = pd.Series(lista, index = [3,4,5,6,7])
display(serie_lista2)

print(" ------ SERIES a partir de LISTAS ESPECIFICANDO ÍNDICE STRING ------ ")

# como hemos mencionado los índices pueden ser de distintos tipos, en este caso crearemos una nueva Serie donde sus índices sean strings
serie_lista3 = pd.Series(lista, index = ["Lunes","Martes","Miercoles","Jueves","Viernes"])
display(serie_lista3)

 ------ SERIES a partir de LISTAS ESPECIFICANDO ÍNDICE NÚMERICO ------ 


3    109
4     78
5      4
6    189
7     98
dtype: int64

 ------ SERIES a partir de LISTAS ESPECIFICANDO ÍNDICE STRING ------ 


Lunes        109
Martes        78
Miercoles      4
Jueves       189
Viernes       98
dtype: int64

In [10]:
print(" ------ SERIES a partir de DICCIONARIOS ------ ")

# definir un diccionario 
dicc = {'Manolo' : 20, 
        'Lorena' : 220, 
        'Paula' : 198, 
        'Lola': 78, 
       'Marta': 167, 
        'Pablo': 48} 

# crear una serie nueva a partir del diccionario definido previamente
serie_diccionario = pd.Series(dicc)
serie_diccionario

 ------ SERIES a partir de DICCIONARIOS ------ 


Manolo     20
Lorena    220
Paula     198
Lola       78
Marta     167
Pablo      48
dtype: int64

## Propiedades de las Series

Las propiedades son atributos específicos que nos proporcionan información sobre las características de una **Serie** en particular. Las propiedades más importantes de las **Series** :

1. `values`: Devuelve los valores de la **Serie** como un *array* NumPy. Podemos acceder a ellos utilizando la sintaxis `serie.values`.

2. `index`: Devuelve los índices de la **Serie**. Como hemos dicho, los índices pueden ser de cualquier tipo, incluyendo *integer*s, *strings* u otros tipos o estructuras de datos. Podemos acceder a ellos utilizando la sintaxis `serie.index`.

3. `dtype`: Devuelve el tipo de datos de los elementos en la Serie. Usaremos la sintaxis `serie.dtype`. Los tipos de datos más comunes que podemos encontrar en una Serie son `int64`, `float64`, `object` (que es el *string* que conocemos de Python básico), `datetime64`, entre otros.

4. `size`: Devuelve el número total de elementos en la Serie. Usaremos la sintaxis `serie.size`.

5. `shape`: Devuelve una tupla que representa la forma (dimensiones) de la **Serie**. En el caso de una **Serie** unidimensional, la forma será `(n,)`, donde `n` es el número de elementos en la Serie. Usaremos la sintaxis `serie.shape`.

Estas propiedades son útiles para obtener información sobre una **Serie** en particular, como el tipo de datos, el número de elementos y los índices asociados. Además, las propiedades pueden ser utilizadas en conjunto con los métodos y funciones de Pandas para realizar operaciones y análisis más avanzados en los datos de la **Serie**.

Es importante tener en cuenta que las **Series** son inmutables, lo que significa que no podemos cambiar sus elementos directamente. Sin embargo, podremos realizar operaciones y manipulaciones para obtener nuevos resultados o modificar la estructura de los datos en función de nuestras necesidades.

In [11]:
# recordemos la serie
display("La serie con la que vamos a trabajar es:", serie_diccionario)
print("---------------------------")

# utilizamos el método index que nos va a devolver todos los índices 
# fijaos como en este caso nos ha devuelto todos los índices de nuestra Serie, es decir, los nombres de las alumnas
print(f"El método 'index' nos devuelve los índices de la Serie: {serie_diccionario.index}")
print("---------------------------")

# ahora utilizamos el método 'values' nos devuelve todos los valores de la Serie
# en este caso los números que teníamos
print(f"El método 'values' nos devuelve los valores de la Serie: {serie_diccionario.values}")
print("---------------------------")

#utilizando el método 'shape' nos devuelve el número de "filas" que tenemos en la Serie
print(f"El método 'shape' nos devuelve la forma de la Serie: {serie_diccionario.shape}")
print("---------------------------")

#utilizando el método 'size' nos devuelve el número de elementos que tenemos en la Serie
print(f"El método 'size' nos devuelve el número de elementos de la Serie: {serie_diccionario.size}")
print("---------------------------")

#utilizando el método 'dtypes' nos devuelve el tipo de datos de la Serie
print(f"El método 'dtypes' nos devuelve el tipo de datos de la Serie: {serie_diccionario.dtypes}")



'La serie con la que vamos a trabajar es:'

Manolo     20
Lorena    220
Paula     198
Lola       78
Marta     167
Pablo      48
dtype: int64

---------------------------
El método 'index' nos devuelve los índices de la Serie: Index(['Manolo', 'Lorena', 'Paula', 'Lola', 'Marta', 'Pablo'], dtype='object')
---------------------------
El método 'values' nos devuelve los valores de la Serie: [ 20 220 198  78 167  48]
---------------------------
El método 'shape' nos devuelve la forma de la Serie: (6,)
---------------------------
El método 'size' nos devuelve el número de elementos de la Serie: 6
---------------------------
El método 'dtypes' nos devuelve el tipo de datos de la Serie: int64


## Indexación de las Series

La indexación se refiere a la forma en que accedemos a los elementos de una **Serie** utilizando índices o etiquetas. Pandas ofrece varias opciones para indexar y seleccionar elementos. Algunas de las principales formas de realizar la indexación en las Series son:

-  **Indexación por posición**: Podemos acceder a los elementos de una **Serie** utilizando la posición numérica del elemento en la Serie, como hacíamos con las listas

-  **Indexación por etiqueta**: Podemos acceder a los elementos de una **Serie** utilizando las etiquetas de índice asociadas a cada elemento. 

-  **Indexación por rango**: Podemos acceder a un rango de elementos de una **Serie** utilizando los índices de inicio y fin. 

- **Indexación por lista de índices**: Podemos acceder a varios elementos de una **Serie** utilizando una lista de índices.


In [12]:
print(" ##### INDEXACIÓN POR POSICIÓN #####")
display("La serie con la que vamos a trabajar es:", serie_diccionario)
print("---------------------------")

# accedemos al primer elemento (10). 🚨 Fijaos como los índices son como los de las listas, empiezan en 0
print("El valor que está en posición 0 es:", serie_diccionario[0])  
print("---------------------------")

# accedemos al tercer elemento (30)
print("El valor que está en posición 2 es:", serie_diccionario[2])  


 ##### INDEXACIÓN POR POSICIÓN #####


'La serie con la que vamos a trabajar es:'

Manolo     20
Lorena    220
Paula     198
Lola       78
Marta     167
Pablo      48
dtype: int64

---------------------------
El valor que está en posición 0 es: 20
---------------------------
El valor que está en posición 2 es: 198


In [13]:
print(" ##### INDEXACIÓN POR ETIQUETA #####")
display("La serie con la que vamos a trabajar es:", serie_diccionario)
print("---------------------------")


# accedemos al valor de la etiqueta de "Lola" (78). 🚨 Esto lo podríamos comparar con los diccionarios
print("El valor que está en posición 0 es:", serie_diccionario["Lola"])  
print("---------------------------")

# accedemos al valor de la etiqueta de "Pablo" (48). 
print("El valor que está en posición 2 es:", serie_diccionario["Pablo"])  

 ##### INDEXACIÓN POR ETIQUETA #####


'La serie con la que vamos a trabajar es:'

Manolo     20
Lorena    220
Paula     198
Lola       78
Marta     167
Pablo      48
dtype: int64

---------------------------
El valor que está en posición 0 es: 78
---------------------------
El valor que está en posición 2 es: 48


In [14]:
print(" ##### INDEXACIÓN POR RANGO #####")
display("La serie con la que vamos a trabajar es:", serie_diccionario)
print("---------------------------")

# accedemos a los elementos desde el índice 1 al 3 ([220, 198, 78]). 🚨 De niuevo fijaos como esto funciona igual que en listas
# nos esta incluyendo el valor que está en el índice 1, pero no el que está en el índice 4
print("Los valores de los índices que están entre las posiciones 1 y 3 (incluida) es: \n", serie_diccionario[1:4])  

 ##### INDEXACIÓN POR RANGO #####


'La serie con la que vamos a trabajar es:'

Manolo     20
Lorena    220
Paula     198
Lola       78
Marta     167
Pablo      48
dtype: int64

---------------------------
Los valores de los índices que están entre las posiciones 1 y 3 (incluida) es: 
 Lorena    220
Paula     198
Lola       78
dtype: int64


In [15]:
print(" ##### INDEXACIÓN POR LISTA DE ÍNDICES #####")
display("La serie con la que vamos a trabajar es:", serie_diccionario)
print("---------------------------")

# accedemos a los elementos con índices 1, 3 y 4. 🚨 Fijaos como en este caso, le hemos puesto dos corchetes
print("Los valores de los índices que están en las posiciones 1,3, y 4 son: \n", serie_diccionario[[0,2,3]])  



 ##### INDEXACIÓN POR LISTA DE ÍNDICES #####


'La serie con la que vamos a trabajar es:'

Manolo     20
Lorena    220
Paula     198
Lola       78
Marta     167
Pablo      48
dtype: int64

---------------------------
Los valores de los índices que están en las posiciones 1,3, y 4 son: 
 Manolo     20
Paula     198
Lola       78
dtype: int64


# DataFrames en Pandas

## ¿Qué es un DataFrame?

Un **DataFrame** es una estructura de datos bidimensional en forma de tabla, similar a una hoja de cálculo o una base de datos relacional. Es una de las estructuras de datos más utilizadas en el análisis de datos con Pandas.

Las principales características de los **DataFrames** son:

1. **Estructura tabular**: Están organizados en filas y columnas. Cada columna representa una variable o característica, y cada fila representa una observación o entrada de datos. Esto facilita la visualización y el manejo de los datos en forma tabular.

2. **Flexibilidad**: Al contrario que las Series, los **DataFrames** pueden contener datos de diferentes tipos, como *integer*s, *float*s, *strings*, fechas, *bool*s, entre otros. Además, las columnas pueden tener nombres personalizados y los índices pueden ser etiquetados con valores personalizados.

3. **Operaciones eficientes**: Los **DataFrames** están diseñados para realizar operaciones eficientes en grandes conjuntos de datos. Ofrecen una amplia gama de funciones y métodos para manipular y analizar datos de manera rápida y eficiente.

4. **Facilidad de importación y exportación**: Los **DataFrames** pueden leer y escribir datos en diversos formatos, como CSV, Excel, archivos de texto, bases de datos SQL, JSON, entre otros. Esto facilita la importación y exportación de datos desde y hacia diferentes fuentes de datos.

5. **Indexación flexible**: Los **DataFrames** permiten acceder y manipular los datos utilizando diversas técnicas de indexación, como indexación por posición, indexación por etiqueta, indexación booleana y más. Esto permite seleccionar y filtrar datos de manera precisa y eficiente.

## Creación de DataFrames

En Pandas, podemos crear **DataFrames** a partir de una variedad de fuentes de datos. Las formas más comunes para crear *DataFrames* son:

1. **A partir de un diccionario de listas**: Podemos crear un **DataFrame** a partir de un diccionario de listas, donde cada clave representa una columna y cada valor es una lista representa los valores de esa columna. 

2. **A partir de una lista de diccionarios**: Podemos crear un **DataFrame** a partir de una lista de diccionarios, donde cada diccionario representa una fila y las claves del diccionario representan las columnas. 


In [16]:
print(" ------ DATAFRAMES a partir de DICCIONARIO ------ ")

# definimos un diccionario que luego convertiremos a DataFrame. Recordemos que esta era la forma que utilizamos en las lecciones del módulo 2 de web scraping
data_diccionario = {
    'Nombre': ['Lola', 'Manolo', 'Pablo', 'María'],
    'Edad': [34, 89, 12, 52],
    'Ciudad': ['Londres', 'Madrid', 'Roma', 'Paris']
}

# convertimos el diccionario en DataFrame usando el método 'pd.DataFrame'
df_diccionario = pd.DataFrame(data_diccionario)

# mostramos el DataFrame. Recordad que para ver las 5 primeras filas usaremos el método '.head()' 
df_diccionario.head()


 ------ DATAFRAMES a partir de DICCIONARIO ------ 


Nombre  Edad   Ciudad
0    Lola    34  Londres
1  Manolo    89   Madrid
2   Pablo    12     Roma
3   María    52    Paris

In [17]:
print(" ------ DATAFRAMES a partir de LISTA ------ ")

# definimos una lista de diccionarios que luego usaremos para convertir en DataFrame
data_lista = [
    {'Nombre': 'Sofía', 'Edad': 22, 'Ciudad': 'México DF'},
    {'Nombre': 'Luis', 'Edad': 27, 'Ciudad': 'Guadalajara'},
    {'Nombre': 'Laura', 'Edad': 33, 'Ciudad': 'Monterrey'},
    {'Nombre': 'Carlos', 'Edad': 29, 'Ciudad': 'Puebla'}
]


# usamos el método 'pd.DataFrame' para convertir la lista de diccionarios a DataFrame
df_lista = pd.DataFrame(data_lista)

# mostramos el DataFrame, en este caso usaremos el método '.tail()' para mostrar las últimas 5 filas. 
df_lista.tail()

 ------ DATAFRAMES a partir de LISTA ------ 


Nombre  Edad       Ciudad
0   Sofía    22    México DF
1    Luis    27  Guadalajara
2   Laura    33    Monterrey
3  Carlos    29       Puebla

## Lectura de archivos en Pandas

En Pandas, disponemos de diversos métodos para abrir y leer diferentes tipos de archivos. Algunos de los métodos más comunes utilizados para la lectura de archivos en Pandas:

- `pd.read_csv`

- `pd.read_excel`

- `pd.read_json`

- `pd.read_pickle`

## `pd.read_csv()`: 

Nos va a permitir leer archivos csv, su sintaxis básica es: 

```python
pd.read_csv(ruta_al_fichero , sep=',', delimiter=None, header='infer', names=None, index_col=None, dtype=None)
```

¿Qué significan todos estos parámetros?

- `ruta_al_fichero`: La dirección de tu archivo CSV que queremos leer.

- `sep` (opcional): Para indicarle a Pandas cómo están separados los datos en nuestro archivo. Normalmente es una coma, pero si es algo diferente, aquí es donde deberíamos cambiarlo.

- `delimiter` (opcional): Es lo mismo que `sep`, solo que con un nombre diferente. Puedes usar cualquiera de los dos.

- `header` (opcional): Indicará la fila del archivo CSV que queremos que sea el encabezado del **DataFrame**. Puede tomar los siguientes valores:
  
  - `None`: No hay encabezado y los datos se leerán sin nombres de columna.
  
  - `'infer'`: Intentará inferir automáticamente los nombres de columna del archivo, este es el que esta por defecto.

  - Un número *integer* que especifica la fila del archivo que se utilizará como encabezado de columna (por ejemplo, `header=0`).
 

- `names` (opcional): Si quieres cambiar los nombres de las columnas por algo más específico. Debe ser una lista.Si se proporciona, reemplazará los nombres inferidos o los nombres de columna del archivo.

  
- `index_col` (opcional): Especifica la columna (o columnas) que se utilizará/n como índice del **DataFrame**. Puede ser un número *integer*, una cadena con el nombre de la columna o una lista de columnas.

- `dtype` (opcional): Esto es para especificar el tipo de datos de tus columnas. 

- `usecols` (opcional): Carga solo columnas especificadas, las columnas deben ser pasadas en formato lista. 
  
  
Estos son solo los conceptos básicos, pero `pd.read_csv()` tiene un montón más de trucos en la manga, como saltarse filas, manejar valores nulos y más. Puedes echar un vistazo a la documentación oficial [aquí](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) para descubrir más sobre todas las cosas geniales que puedes hacer con esta función.

In [19]:
df_jobs = pd.read_csv("../Datos/Pandas/aire.csv")
df_jobs.head(1)

FileNotFoundError: [Errno 2] No such file or directory: '../Datos/Pandas/aire.csv'

Como la propia definición indica, nuestras columnas deben ir separadas por comas. A veces, puede ocurrir que no vengan separados por comas. Como en el csv que acabamos de abrir. Y es que si nos fijamos nuestras columnas están separadas por `;`. ¿Cómo podemos solucionar esto? 

> Usando el parámetro `sep` de `pd.read_csv` donde tendremos que especificar por qué están separadas mis columnas.

In [14]:
df_aire = pd.read_csv("../Datos/Pandas/aire.csv", sep = ";")
df_aire.head(1)

provincia  municipio  estacion  magnitud punto_muestreo   ano  mes  dia  \
0         28        102         1         1  28102001_1_38  2022    1   30   

   h01 v01  h02 v02  h03 v03  h04 v04  h05 v05  h06 v06  h07 v07  h08 v08  \
0  3.0   T  3.0   T  3.0   T  3.0   T  3.0   T  3.0   T  3.0   T  3.0   T   

   h09 v09  h10 v10  h11 v11  h12 v12  h13 v13  h14 v14  h15 v15  h16 v16  \
0  3.0   T  3.0   T  3.0   T  3.0   T  NaN   N  NaN   N  NaN   N  NaN   N   

   h17 v17  h18 v18  h19 v19  h20 v20  h21 v21  h22 v22  h23 v23  h24 v24  
0  NaN   N  NaN   N  NaN   N  NaN   N  NaN   N  NaN   N  NaN   N  NaN   N

## `pd.read_excel()`

La sintaxis básica de `pd.read_excel` es la siguiente:

```python
pd.read_excel(ruta_al_fichero, sheet_name=0, header=0, names=None, index_col=None, dtype=None)
```

Donde:
- `ruta_al_fichero`:  Es el nombre del archivo o la ruta de Excel que queremos leer. 

- `sheet_name` (opcional): Indica la hoja o hojas del archivo Excel que queremos leer. Puede tener los siguientes valores:

  - Un número *integer* que especifica el índice de la hoja (por ejemplo, `sheet_name=0` para la primera hoja).

  - Un nombre de hoja como una cadena (por ejemplo, `sheet_name='Sheet1'`).

  - Una lista de nombres de hojas para leer varias hojas (por ejemplo, `sheet_name=['Sheet1', 'Sheet2']`).

- `header` (opcional): Indica la fila del archivo Excel que se utilizará como encabezado de columnas. Puede tener los siguientes valores:

  - `None`: No hay encabezado y los datos se leerán sin nombres de columna.

  - Un número *integer* que especifica la fila del archivo que se utilizará como encabezado de columna (por ejemplo, `header=0`).

- `names` (opcional): Es una lista opcional de nombres para las columnas. Si se proporciona, reemplazará los nombres inferidos o los nombres de columna del archivo.

- `index_col` (opcional): Especifica la columna (o columnas) que se utilizará como índice del DataFrame resultante. Puede ser un número *integer*, una cadena con el nombre de la columna o una lista de columnas.

- `dtype` (opcional): Permite especificar el tipo de datos de las columnas. Puede ser un diccionario que mapee nombres de columna a tipos de datos o un solo tipo de datos para aplicar a todas las columnas.

Además de estos parámetros, `pd.read_excel` también admite otros parámetros opcionales, como `skiprows`, `na_values`, `usecols` y más. Estos parámetros nos permiten personalizar aún más la lectura de archivos Excel según tus necesidades. Al igual que en el método `pd.read_csv()`, podemos  consultar la [documentación oficial](https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html) para obtener información más detallada sobre estos parámetros y sus opciones.

In [ ]:
df_xlsx = pd.read_excel("../Datos/Pandas/espacios_protegidos.xlsx", sheet_name=0)
df_xlsx.head()

FileNotFoundError: [Errno 2] No such file or directory: '../Datos/Pandas/espacios_protegidos.xlsx'

🚨⚠️ **NOTA** Puede que al ejecutar este línea de código os salga el siguiente error: 

![error](imagenes/error_excel.png)

**DON'T PANIC!!!** Lo único que tendremos que hacer es importaros `openpyxl`. ¿Cómo? Nos vamos a la terminal y ejecutamos el siguiente código: 

```bash 
pip3 install openpyxl

pip install openpyxl
```

In [16]:
# volvemos a ejecutar el código para abrir el fichero
df_xlsx = pd.read_excel("../Datos/Pandas/espacios_protegidos.xlsx", sheet_name=0)
df_xlsx.head()

espacio_prot_categoria  \
0  Áreas protegidas por instrumentos internacionales   
1  Áreas protegidas por instrumentos internacionales   
2  Áreas protegidas por instrumentos internacionales   
3                Espacios Protegidos Red Natura 2000   
4                Espacios Protegidos Red Natura 2000   

                             espacio_prot_figura  \
0                         Reserva de la Biosfera   
1                         Reserva de la Biosfera   
2  Humedal de Importancia Internacional (RAMSAR)   
3      Zona de Especial Protección para las Aves   
4      Zona de Especial Protección para las Aves   

                                 espacio_prot_nombre  \
0  Cuencas Altas de los ríos Manzanares, Lozoya y...   
1                                  Sierra del Rincón   
2                  Humedales del Macizo de Peñalara    
3                                  Monte de El Pardo   
4                                   Soto de Viñuelas   

   espacio_prot_superficie_ha espacio_prot_fecha_declaracion  \
0                  105655.000                     1992-11-09   
1                   15231.000                     2005-06-29   
2                     487.198                     2005-12-16   
3                   15298.670                     1988-02-01   
4                    3071.890                     1988-02-01   

                              espacio_prot_normativa  \
0  Normativa Áreas protegidas por instrumentos in...   
1  Normativa Áreas protegidas por instrumentos in...   
2  Normativa Áreas protegidas por instrumentos in...   
3      Normativa Espacios Protegidos Red Natura 2000   
4      Normativa Espacios Protegidos Red Natura 2000   

                        espacio_prot_informacion_web  
0  Reserva de la Biosfera Cuencas Altas de los rí...  
1           Reserva de la Biosfera Sierra del Rincón  
2                                   Humedales Ramsar  
3                Espacios protegidos Red Natura 2000  
4                Espacios protegidos Red Natura 2000

## `pd.read_json()`

Nos va a permitir abrir arcivos de tipo json. La sintaxis básica de `pd.read_json` es la siguiente:

```python
pd.read_json(ruta_al_fichero, orient=None, typ='frame', dtype=True, convert_axes=True, convert_dates=True, keep_default_dates=True, numpy=False, precise_float=False, date_unit=None, encoding=None, lines=False)
```

Los parámetros que puede recibir este método son:

- `ruta_al_fichero`: Es la ruta del archivo JSON que deseamos leer.

- `orient` (opcional): Especifica la orientación del archivo JSON. Puede tener los siguientes valores:

  - `'columns'` (por defecto): Se espera que el archivo JSON tenga un formato de columna.

  - `'index'`: Se espera que el archivo JSON tenga un formato de índice.

  - `'records'`: Se espera que el archivo JSON tenga un formato de registros.

  - `'split'`: Se espera que el archivo JSON tenga un formato dividido.

  - `'values'`: Se espera que el archivo JSON contenga solo valores sin ninguna etiqueta de columna o índice.

- `typ` (opcional): Indica el tipo de objeto a crear. Puede tener los siguientes valores:

  - `'frame'` (por defecto): Crea un objeto DataFrame.

  - `'series'`: Crea un objeto Series.

- `dtype` (opcional): Permite especificar el tipo de datos de las columnas. Puede ser un diccionario que mapee nombres de columna a tipos de datos o un solo tipo de datos para aplicar a todas las columnas.

- `convert_axes` (opcional): Indica si las etiquetas de los ejes deben convertirse en índices o nombres de columna. Por defecto, es `True`.

- `convert_dates` (opcional): Indica si se deben convertir las cadenas de fecha y hora en objetos de fecha y hora. Por defecto, es `True`.

- `keep_default_dates` (opcional): Especifica si se deben mantener las fechas predeterminadas al convertir cadenas de fecha y hora. Por defecto, es `True`.

- `numpy` (opcional): Indica si los datos deben devolverse como una matriz NumPy en lugar de un objeto DataFrame. Por defecto, es `False`.

- `precise_float` (opcional): Indica si se deben utilizar números de punto *float* precisos en lugar de valores de punto *float* nativos de Python. Por defecto, es `False`.

- `date_unit` (opcional): Especifica la unidad de fecha y hora si se deben convertir las cadenas de fecha y hora. Puede ser `'s'` para segundos o `'ms'` para milisegundos.

- `encoding` (opcional): Permite especificar la codificación del archivo JSON si no se puede inferir automáticamente.

- `lines` (opcional): Indica si el archivo JSON contiene múltiples objetos JSON en líneas separadas en lugar de un solo objeto JSON.

Estos son solo algunos de los parámetros más comunes utilizados en `pd.read_json`. La función `pd.read_json` también admite otros parámetros opcionales, como `orient`, `date_format`, `numpy`, `compression` y más. Podemos consultar la [documentación oficial](https://pandas.pydata.org/docs/reference/api/pandas.read_json.html) para obtener información más detallada sobre estos parámetros y sus opciones.

In [17]:
df_residuos = pd.read_json("../Datos/Pandas/residuos.json")
df_residuos.head()

data
0  {'residuos_pelig_cantidad_ton': 23476.42, 'res...
1  {'residuos_pelig_cantidad_ton': 1927.25, 'resi...
2  {'residuos_pelig_cantidad_ton': 25333.54, 'res...
3  {'residuos_pelig_cantidad_ton': 22060.07, 'res...
4  {'residuos_pelig_cantidad_ton': 13109.74, 'res...

In [18]:
# esto nos devuelve un DataFrame que esta lleno de diccionarios!
# sin embargo, lo podemos convertir fácilmente a algo más legible usando la siguiente sintaxis

df_residuos2 = df_residuos['data'].apply(pd.Series)
df_residuos2.head()

residuos_pelig_cantidad_ton  residuos_pelig_año  \
0                     23476.42                2012   
1                      1927.25                2012   
2                     25333.54                2012   
3                     22060.07                2012   
4                     13109.74                2012   

                       residuos_pelig_opcion_gestion  \
0                                          Reciclado   
1                                          Reciclado   
2                                          Reciclado   
3  Tratamiento previo a otras formas de valorización   
4  Tratamiento previo a otras formas de valorización   

                      residuos_pelig_tratamiento  
0                    Recuperación de disolventes  
1                        Recuperación de metales  
2                         Regeneración de aceite  
3  Trituración previa a valorización de baterías  
4     Operaciones previas a valorización de RAEE

## `pd.read_pickle`


**¿Qué es Pickle exactamente?**

Un archivo de tipo pickle es un archivo binario en Python que se utiliza para serializar y deserializar objetos,  su acrónimo viene de *Python Pickle Format* que fue desarrollado exclusivamente para Python. Leyendo solo esto podemos seguir sin tener claro que es un archivo de tipo pickle. Expliquemoslo de una forma más sencilla. Imagina que tienes un montón de objetos especiales en tu habitación, como juguetes, libros, peluches, etc. Ahora, supongamos que quieres guardar todos estos objetos en una caja para poder llevarlos contigo o almacenarlos de manera segura.

Aquí es donde entra en juego un archivo pickle. Es como una caja especial que te permite guardar todos estos objetos en ella. Pero, en lugar de objetos físicos, el archivo pickle puede almacenar objetos digitales, como datos, números, listas, diccionarios y más.

El archivo pickle es como una caja mágica que puede guardar todos tus objetos digitales de Python. Cuando guardas estos objetos en un archivo pickle, se convierten en una forma especial que se puede almacenar en tu computadora o enviar a alguien más.

La sintaxis básica de `pd.read_pickle` es la siguiente:

```python
pd.read_pickle(ruta_al_fichero, compression='infer')
```

Donde:

- `filepath_or_buffer`: Es la ruta del archivo pickle que deseas leer.

- `compression` (opcional): Es un parámetro opcional que indica el tipo de compresión utilizado en el archivo pickle. Por defecto, es `'infer'`, lo que significa que la biblioteca intentará inferir automáticamente el tipo de compresión. Puedes especificar un tipo de compresión explícitamente, como `'gzip'` o `'bz2'`, si el archivo pickle está comprimido.


In [19]:
df_pkl = pd.read_pickle("../Datos/Pandas/paises_capitales.pkl")
df_pkl

{'Colombia': 'Bogotá', 'Ecuador': 'Quito', 'Argentina': 'Buenos'}

## Indexación en Pandas (`loc` e `iloc`)

En Pandas, tanto `loc` como `iloc` son métodos para acceder y manipular datos en un **DataFrame**. Ambos métodos ofrecen formas de indexar datos, ya sea utilizando etiquetas o números, respectivamente.

- `loc`: Este método se utiliza para acceder a los datos utilizando etiquetas de fila y columna. Con `loc`, podemos seleccionar filas y columnas **utilizando sus nombres o etiquetas**. La sintaxis general para `loc` es `df.loc[nombre_filas, nombre_columnas]`, donde `filas` y `columnas` pueden ser un único nombre o una lista de nombres.

- `iloc`: Se utiliza para acceder a los datos utilizando índices de filas y columnas. Con `iloc`, podemos seleccionar filas y columnas **basándonos en sus posiciones numéricas**. La sintaxis general para `iloc` es `df.iloc[índice_filas, índice_columnas]`, donde `filas` y `columnas` pueden ser un único número(índice) o una lista de números(índices).

La principal diferencia entre `loc` y `iloc` radica en cómo se realizan las selecciones:

- `loc`: Utiliza etiquetas o nombres de filas y columnas para indexar los datos. Por ejemplo, `df.loc[3, 'A']` seleccionará el valor en la fila con etiqueta 3 y la columna con etiqueta 'A'.

- `iloc`: Por otro lado, `iloc` utiliza índices basados en la posición para realizar las selecciones. Por ejemplo, `df.iloc[3, 0]` seleccionará el valor en la cuarta fila y primera columna (recordemos que los índices empiezan en 0).

Además, debemos tener en cuenta que `loc` incluye el límite superior en los rangos de selección, mientras que `iloc` lo excluye. Por ejemplo, `df.loc[1:3, 'A']` seleccionará las filas con etiquetas 1, 2 y 3, mientras que `df.iloc[1:3, 0]` seleccionará las filas con índices 1 y 2 (excluyendo el índice 3).


In [24]:
# importamos un nuevo fichero de datos para entender mejor estos conceptos
df = pd.read_csv("../Datos/Pandas/medallas.csv",index_col=["País"],  usecols=["Oro","Plata", "Bronce", "País"])
df

Oro  Plata  Bronce
País                                
China            1473    994     720
Japón            1032   1037     985
Corea del Sur     745    663     827
Irán              179    181     197
Kazajistán        155    158     224
India             154    202     315
Tailandia         132    175     278
Indonesia         112    131     240
China Taipéi       99    144     276
Corea del Norte    91    120     235

### Selección a través de un único elemento

Vamos a intentar sacar el valor de 1037, en concreto hace referencia a las medallas de Plata de Japón. Recordemos que: 

- En el `loc` usaremos los nombres de las filas y las columnas (es como el juego de hundir la flota), corresponde a lo que tenemos en rojo en la imagen. 

- En el `iloc` usaremos los indices de las filas y las columnas, igual que hacíamos en las listas o los **array*s* de NumPy, corresponde con lo que vemos en verde en la imagen. 

![imagen](https://github.com/Hack-io-Data/Imagenes/blob/main/02-Imagenes/Pandas/dataframe_medallas_plata_japon.png?raw=true)

In [27]:
# Usando el loc, accediendo por los nombres de las filas y las columnas
print(f'El número de medallas de Plata conseguidas por Japón usando el LOC son:  {df.loc["Japón", "Plata"]}')

# usando el iloc, accediendo por los índices, que empiezan en 0 como siempre en Python
print(f'El número de medallas de Plata conseguidas por Japón usando el ILOC es:  {df.iloc[1, 1]}')

El número de medallas de Plata conseguidas por Japón usando el LOC son:  1037
El número de medallas de Plata conseguidas por Japón usando el ILOC es:  1037


Imaginemos que queremos sacar una fila en concreto pero todas las columnas o viceversa:

> En estos casos tendremos que usar `:`, esto índica que lo queremos todo. Es decir: 

```python
[filas_deseadas, :] # le estamos diciendo que nos busque por una fila dada y nos devuelva todas las columnas

[:, columna_deseada] # devuelvenos todas las filas de la columna que especifiquemos
```

Imaginemos ahora que queremos extraer toda la información de Japón. En este caso, querremos de una fila concreta y todas las columnas. 

![imagen2](https://github.com/Hack-io-Data/Imagenes/blob/main/02-Imagenes/Pandas/dataframe_medallas_plata_japon.png?raw=true)

In [28]:
# Usando el loc, accediendo por los nombres de las filas y las columnas
print(f'El número de medallas conseguidas por Japón  usando el LOC es: \n  {df.loc["Japón", :]}')

print("\n----------------------\n")
# Usando el iloc, accediendo por los índices
print(f'El número de medallas conseguidas por Japón usando el ILOC es: \n  {df.iloc[1, :]}')

El número de medallas conseguidas por Japón  usando el LOC es: 
  Oro       1032
Plata     1037
Bronce     985
Name: Japón, dtype: int64

----------------------

El número de medallas conseguidas por Japón usando el ILOC es: 
  Oro       1032
Plata     1037
Bronce     985
Name: Japón, dtype: int64


¿Y si ahora quisieramos extraer todas las medallas de Oro para todos los países? En este caso querríamos todas las filas y una sola columna. 

![imagen3](https://github.com/Hack-io-Data/Imagenes/blob/main/02-Imagenes/Pandas/dataframe_medallas_oro.png?raw=true)

In [29]:
# Usando el loc, accediendo por los nombres de las filas y las columnas
print(f'El número de medalas de Oro de todos los países usando el LOC es: \n  {df.loc[:, "Oro"]}')

print("\n----------------------\n")
# Usando el iloc, accediendo por los índices
print(f'El número de medalas de Oro de todos los países usando el ILOC es: \n  {df.iloc[:, 0]}')

El número de medalas de Oro de todos los países usando el LOC es: 
  País
China              1473
Japón              1032
Corea del Sur       745
Irán                179
Kazajistán          155
India               154
Tailandia           132
Indonesia           112
China Taipéi         99
Corea del Norte      91
Name: Oro, dtype: int64

----------------------

El número de medalas de Oro de todos los países usando el ILOC es: 
  País
China              1473
Japón              1032
Corea del Sur       745
Irán                179
Kazajistán          155
India               154
Tailandia           132
Indonesia           112
China Taipéi         99
Corea del Norte      91
Name: Oro, dtype: int64


### Selección a través de una lista de valores 

En este caso la sintaxis del `loc` y el `iloc` cambian un poco. Seguiremos usando los nombres de las filas y las columnas para el `loc` y los índices para el `iloc`, pero tendremos que añadir un corchete más. 

```python
[[fila_deseada1, fila_deseada2], :] # le estamos diciendo que nos busque por una fila dada y nos devuelva todas las columnas. FIJAOS COMO LO QUE HEMOS HECHO HA SIDO AÑADIR UNOS CORCHETES MÁS DONDE INCLUIREMOS LOS NOMBRES DE LAS FILAS O COLUMNAS QUE QUERRAMOS SEPARADAS POR COMAS. Los ':' siguen indicando que queremo todas las columnas en este caso.

[:, [col_deseada1, col_deseada2]] # devuelvenos todas las filas de la columna que especifiquemos. FIJAOS COMO LO QUE HEMOS HECHO HA SIDO AÑADIR UNOS CORCHETES MÁS DONDE INCLUIREMOS LOS NOMBRES DE LAS FILAS O COLUMNAS QUE QUERRAMOS SEPARADAS POR COMAS. Los ':' siguen indicando que queremos todas las filas.
```

Ahora vamos a querer extraer la información de las medallas de Oro, pero solo vamos a querer los datos de Corea (Corea del Sur y Corea del Norte). 

In [30]:
df

Oro  Plata  Bronce
País                                
China            1473    994     720
Japón            1032   1037     985
Corea del Sur     745    663     827
Irán              179    181     197
Kazajistán        155    158     224
India             154    202     315
Tailandia         132    175     278
Indonesia         112    131     240
China Taipéi       99    144     276
Corea del Norte    91    120     235

In [31]:
# Usando el loc, accediendo por los nombres de las filas y las columnas
print(f'Las medallas de Oro para Corea del Norte y del Sur usando el LOC es: \n  {df.loc[["Corea del Sur", "Corea del Norte"], "Oro"]}')

print("\n----------------------\n")
# Usando el iloc, accediendo por los índices
print(f'Las medallas de Oro para Corea del Norte y del Sur usando el ILOC es: \n  {df.iloc[[2, -1], 0]}')

Las medallas de Oro para Corea del Norte y del Sur usando el LOC es: 
  País
Corea del Sur      745
Corea del Norte     91
Name: Oro, dtype: int64

----------------------

Las medallas de Oro para Corea del Norte y del Sur usando el ILOC es: 
  País
Corea del Sur      745
Corea del Norte     91
Name: Oro, dtype: int64


Imaginemos ahora que no queremos solo la información de las medallas de Oro para Corea del Norte y del Sur, si no que también queremos la información de las medallas de Plata. 

In [33]:
# Usando el loc, accediendo por los nombres de las filas y las columnas
print(f'Las medallas de Oro y Plata para Corea del Norte y del Sur usando el LOC es: \n  {df.loc[["Corea del Sur", "Corea del Norte"], ["Oro", "Plata"]]}')

print("\n----------------------\n")
# Usando el iloc, accediendo por los índices
print(f'Las medallas de Oro y Plata para Corea del Norte y del Sur usando el ILOC es: \n  {df.iloc[[2, -1], [0,1]]}')

Las medallas de Oro y Plata para Corea del Norte y del Sur usando el LOC es: 
                   Oro  Plata
País                       
Corea del Sur    745    663
Corea del Norte   91    120

----------------------

Las medallas de Oro y Plata para Corea del Norte y del Sur usando el ILOC es: 
                   Oro  Plata
País                       
Corea del Sur    745    663
Corea del Norte   91    120


### Selección de un rango de datos a través de un corte

En este caso podremos usar la sintaxis de `[start:stop:step]` que conocemos de listas. 

In [34]:
# lo primero que vamos a hacer es recordar el DataFrame
df

Oro  Plata  Bronce
País                                
China            1473    994     720
Japón            1032   1037     985
Corea del Sur     745    663     827
Irán              179    181     197
Kazajistán        155    158     224
India             154    202     315
Tailandia         132    175     278
Indonesia         112    131     240
China Taipéi       99    144     276
Corea del Norte    91    120     235

En este ejercicio vamos a querer sacar las medallas de Oro y Bronce desde la Irán hasta Indonesia. 

In [36]:
# Usando el loc, accediendo por los nombres de las filas y las columnas
print(f"Las medallas de Oro y Bronce de Irán hasta Indonesia usando el LOC son: \n  {df.loc['Irán': 'Indonesia', ["Oro", "Bronce"] ]}")

print("\n----------------------\n")
# Usando el iloc, accediendo por los índices
print(f'Las medallas de Oro y Bronce de Irán hasta Indonesia usando el ILOC son: \n  {df.iloc[3:8, [0,2]]}')

Las medallas de Oro y Bronce de Irán hasta Indonesia usando el LOC son: 
              Oro  Bronce
País                   
Irán        179     197
Kazajistán  155     224
India       154     315
Tailandia   132     278
Indonesia   112     240

----------------------

Las medallas de Oro y Bronce de Irán hasta Indonesia usando el ILOC son: 
              Oro  Bronce
País                   
Irán        179     197
Kazajistán  155     224
India       154     315
Tailandia   132     278
Indonesia   112     240


¿Y si quisieramos saber todas las medallas de China, Corea del Sur, Kazajistán y Tailandia? En este caso tendremos que hacer uso del `[start_ stop: step]`

In [37]:
df

Oro  Plata  Bronce
País                                
China            1473    994     720
Japón            1032   1037     985
Corea del Sur     745    663     827
Irán              179    181     197
Kazajistán        155    158     224
India             154    202     315
Tailandia         132    175     278
Indonesia         112    131     240
China Taipéi       99    144     276
Corea del Norte    91    120     235

In [47]:
# Usando el loc, accediendo por los nombres de las filas y las columnas
print(f"Todas las medallas para China, Corea del Sur,  Kazajistán y Tailandia usando el LOC es: \n  {df.loc["China":"Tailandia":2, :]}")

print("\n----------------------\n")
# Usando el iloc, accediendo por los índices
print(f'Todas las medallas para China, Corea del Sur,  Kazajistán y Tailandia usando el ILOC es: \n  {df.iloc[:7:2, :]}')

Todas las medallas para China, Corea del Sur,  Kazajistán y Tailandia usando el LOC es: 
                  Oro  Plata  Bronce
País                              
China          1473    994     720
Corea del Sur   745    663     827
Kazajistán      155    158     224
Tailandia       132    175     278

----------------------

Todas las medallas para China, Corea del Sur,  Kazajistán y Tailandia usando el ILOC es: 
                  Oro  Plata  Bronce
País                              
China          1473    994     720
Corea del Sur   745    663     827
Kazajistán      155    158     224
Tailandia       132    175     278


### Seleccionar basado en una condición 

Al igual que en NumPy podemos filtrar los datos de nuestro **DataFrame** usando operadores de comparación,  como `>`, `<`, `>=`, `<=`, `==` o `!=`. Pero antes de ponernos a ver como filtrar datos, aprendamos a acceder solo a una columna de nuestro **DataFrame**. Podríamos pensar que lo podemos hacer con un `loc` o un `iloc` como hemos estado viendo hasta ahora, pero hay una forma un poco más sencilla. Para eso deberemos seguir la siguiente sintaxis:

```python
df["nombre_columna"]
df.nombre_columna
# estas dos opciones me devuelven exactamente lo mismo. 
```

In [48]:
# recordamos el DataFrame
df

Oro  Plata  Bronce
País                                
China            1473    994     720
Japón            1032   1037     985
Corea del Sur     745    663     827
Irán              179    181     197
Kazajistán        155    158     224
India             154    202     315
Tailandia         132    175     278
Indonesia         112    131     240
China Taipéi       99    144     276
Corea del Norte    91    120     235

In [50]:
# accedemos a los datos de la columna 'Oro'
# Fijaos como nos ha devuelto una serie donde tenemos todos los valores de la columna 'Oro'
df["Oro"]

País
China              1473
Japón              1032
Corea del Sur       745
Irán                179
Kazajistán          155
India               154
Tailandia           132
Indonesia           112
China Taipéi         99
Corea del Norte      91
Name: Oro, dtype: int64

In [51]:
# El código de arriba es exactamente igual que:
df.Oro

País
China              1473
Japón              1032
Corea del Sur       745
Irán                179
Kazajistán          155
India               154
Tailandia           132
Indonesia           112
China Taipéi         99
Corea del Norte      91
Name: Oro, dtype: int64

¿Y que pasaría si quisiera acceder a varias columnas? Solo podremos usar la sintaxis `df[["columna_1", "columna_2"]]`. Fijaos como ahora estamos añadiendo un corchete extra a la sintaxis que hemos aprendido en los pasos anteriores. 

In [52]:
df[["Oro", "Bronce"]]

Oro  Bronce
País                         
China            1473     720
Japón            1032     985
Corea del Sur     745     827
Irán              179     197
Kazajistán        155     224
India             154     315
Tailandia         132     278
Indonesia         112     240
China Taipéi       99     276
Corea del Norte    91     235

Una vez que hemos visto esto como acceder a los datos de una o varias columnas unicamente, vamos a ver como podemos filtrar los datos del **DataFrame**, igual que hacíamos en los *array*s. 

En este primer ejemplo, vamos a quedarnos con todos los datos donde las medallas de Oro  sean mayor que 1000. En este caso seguiremos la siguiente sintaxis: 

```python
df.loc[df["columna"] CONDICION]
```

In [53]:
# filtramos nuestros datos para quedarnos solo con las filas donde la temperatura sean mayor que 12
df_mayor_1000_1 = df.loc[df.Oro > 1000, :]
print(f"El resultado usando la sintaxis de DF.COLUMNA usando LOC es:\n {df_mayor_1000_1}")

print("\n --------------- \n")

# la linea de arriba es exactamente la misma que esta que vemos aquí abajo
df_mayor_1000_2 = df.loc[df["Oro"] > 1000, :]
print(f"El resultado usando la sintaxis de df['COLUMNA'] usando LOC es:\n {df_mayor_1000_2}")


El resultado usando la sintaxis de DF.COLUMNA usando LOC es:
         Oro  Plata  Bronce
País                      
China  1473    994     720
Japón  1032   1037     985

 --------------- 

El resultado usando la sintaxis de df['COLUMNA'] usando LOC es:
         Oro  Plata  Bronce
País                      
China  1473    994     720
Japón  1032   1037     985


In [54]:
# ¿y como lo haríamos con iloc? Se nos ocurriría que usando los corchetes, poniendo el índice de la columna que queremos usar como filtro. 
# veamoslo. 
df.iloc[df[0] > 1000, :]

KeyError: 0

⚠️ Obtenemos el error porque `iloc` no puede aceptar una Serie booleana. Sólo acepta una lista booleana. Podemos utilizar la función `list()` para convertir una Serie en una lista booleana. Por lo tanto, tendríamos que:

In [55]:
df.iloc[list(df.Oro > 1000), :]

Oro  Plata  Bronce
País                      
China  1473    994     720
Japón  1032   1037     985

In [56]:
# si juntamos todo lo aprendido de loc e iloc en las últimas celdas

# filtramos nuestros datos para quedarnos solo con las filas donde el número de medallas de Oro sea mayor que 1000
df_mayor_1000_1 = df.loc[df.Oro > 1000, :]
print(f"El resultado usando la sintaxis de DF.COLUMNA usando LOC es:\n {df_mayor_1000_1}")

print("\n --------------- \n")

# la linea de arriba es exactamente la misma que esta que vemos aquí abajo
df_mayor_1000_2 = df.loc[df["Oro"] > 1000, :]
print(f"El resultado usando la sintaxis de df['COLUMNA'] usando LOC es:\n {df_mayor_1000_2}")

print("\n --------------- \n")

# filtramos con ILOC
df_mayor_1000_iloc = df.iloc[list(df.Oro > 1000), :]
print(f"El resultado usando la sintaxis de df['COLUMNA'] usando ILOC es:\n {df_mayor_1000_iloc}")



El resultado usando la sintaxis de DF.COLUMNA usando LOC es:
         Oro  Plata  Bronce
País                      
China  1473    994     720
Japón  1032   1037     985

 --------------- 

El resultado usando la sintaxis de df['COLUMNA'] usando LOC es:
         Oro  Plata  Bronce
País                      
China  1473    994     720
Japón  1032   1037     985

 --------------- 

El resultado usando la sintaxis de df['COLUMNA'] usando ILOC es:
         Oro  Plata  Bronce
País                      
China  1473    994     720
Japón  1032   1037     985


**¿Y si queremos que se cumplan varias condiciones?**

En este caso tendremos que usar los operadores `&` o `|`, como vimos en NumPy. Recordemos lo que significan: 

- `&`: es como el operador `and`. En este caso se tendrán que cumplir todas las condiciones que le pasemos. 

- `|`: es como el operador `or`. En este caso nos devolverá datos de las condiciones que se cumplan, pero tienen porque cumplirse todas a la vez. O una condición y otra. 

En este ejercicio vamos a querer filtrar  los países que tengan menos de 100 medallas de Oro  y más de 900 de Bronce. 

In [57]:
# Seleccionar aquellas filas donde se cumpla la condición de que el viento es mayor que 20 y el tiempo era soleado.
# además estamos especificando que nos devuelva solo las columnas de "Temperature" y " Wind"
# SE TIENEN QUE CUMPLIR LAS DOS CONDICIONES

# filtramos con LOC
df_dos_condiciones_loc = df.loc[(df.Oro < 100) & (df.Bronce > 900), :]

# filtramos con ILOC
df_dos_condiciones_iloc = df.iloc[list((df.Oro < 100) & (df.Bronce > 900)), :]


print(f"El resultado usando LOC es:\n {df_dos_condiciones_loc}")

print("\n --------------- \n")

print(f"El resultado usando ILOC es:\n {df_dos_condiciones_iloc}")

El resultado usando LOC es:
 Empty DataFrame
Columns: [Oro, Plata, Bronce]
Index: []

 --------------- 

El resultado usando ILOC es:
 Empty DataFrame
Columns: [Oro, Plata, Bronce]
Index: []


In [58]:
# hagámoslo ahora con un or (|)
# SE TIENE QUE CUMPLIR UNA CONDICIÓN U OTRA

# filtramos con LOC
df_dos_condiciones_loc_or = df.loc[(df.Oro < 100) | (df.Bronce > 900), :]

# filtramos con ILOC
df_dos_condiciones_iloc_or = df.iloc[list((df.Oro < 100) | (df.Bronce > 900)), :]


print(f"El resultado usando LOC es:\n {df_dos_condiciones_loc_or}")

print("\n --------------- \n")

print(f"El resultado usando ILOC es:\n {df_dos_condiciones_iloc_or}")


El resultado usando LOC es:
                   Oro  Plata  Bronce
País                                
Japón            1032   1037     985
China Taipéi       99    144     276
Corea del Norte    91    120     235

 --------------- 

El resultado usando ILOC es:
                   Oro  Plata  Bronce
País                                
Japón            1032   1037     985
China Taipéi       99    144     276
Corea del Norte    91    120     235


**POR LO TANTO, LAS DIFERENCIAS ENTRE `LOC` E `ILOC` SON:**

|       | loc                                             | iloc                                                |
|-------|-------------------------------------------------|-----------------------------------------------------|
| Uso   | Acceso a datos utilizando etiquetas de índice    | Acceso a datos utilizando posiciones de índice       |
| Índice| Puede utilizar etiquetas o slices de etiquetas   | Solo puede utilizar posiciones enteras o slices      |
| Filtrado | Permite filtrar filas y seleccionar columnas     | Permite filtrar filas y seleccionar columnas         |
| Sintaxis | Utiliza etiquetas separadas por comas           | Utiliza posiciones enteras separadas por comas       |
| Ejemplo | `df.loc[3:5, 'columna']`                        | `df.iloc[3:5, 2]`                                   |
